File: Part2/Selection_Sort/selection_sort.py

input = [5, 6, 4, 2, 1, 3]


def selection_sort(list_data):
    count = 0
    for i in range(0, (len(list_data) - 1)):
        position = i
        minimum = list_data[i]
        for j in range(i, (len(list_data))):
            if j == i:
                continue
            if list_data[j] < minimum:
                minimum = list_data[j]
                position = j
        if position == i:
            continue
        tmp = list_data[i]
        list_data[i] = minimum
        list_data[position] = tmp
        count += 1
    return count, list_data


if __name__ == "__main__":
    count, data = selection_sort(input)
    print(count)
    print(*data)

File: Part2/Selection_Sort/selection_sort.c
#include <stdio.h>

int count = 0;
void displayList(int *list, int length)
{
    for (int i = 0; i < length; i++)
    {
        printf("%d ", list[i]);
    }
    printf("\n");
}

int *selectionSort(int *list, int length)
{
    for (int i = 0; i < length; i++)
    {
        int mini = i;
        for (int j = i; j < length; j++)
        {
            if (list[j] < list[mini])
            {
                mini = j;
            }
        }
        if (mini != i)
        {
            int tmp = list[i];
            list[i] = list[mini];
            list[mini] = tmp;
            count += 1;
        }
    }
    return list;
}

int main()
{
    printf("要素数を入力\n");
    int length;
    scanf("%d", &length);
    int input[length];

    printf("値を入力\n");
    for (int a = 0; a < length; a++)
    {
        scanf("%d", &input[a]);
    }

    int *result = selectionSort(input, length);
    displayList(result, length);
    printf("%d\n", count);
    return 0;
}